### PR TITLE
separate entry point for each user interface

### PR DESCRIPTION
**Pull request by itswisdomagain**
- root `main.go` and associated `InterfaceMode` config option (--mode) are removed
- new `./cmd` package to hold separate entry points for each interface
- readme updated accordingly

**Using go run**
```
go run ./cmd/godcr-cli
go run ./cmd/godcr-fyne
go run ./cmd/godcr-nuklear
go run ./cmd/godcr-terminal
go run ./cmd/godcr-web
```

**Using go install**
```
go install ./cmd/...
```
or any of the following to install specific interface binaries
```
go install ./cmd/godcr-cli
go install ./cmd/godcr-fyne
go install ./cmd/godcr-nuklear
go install ./cmd/godcr-terminal
go install ./cmd/godcr-web
```

**Using go build**
```
go build ./cmd/godcr-cli
go build ./cmd/godcr-fyne
go build ./cmd/godcr-nuklear
go build ./cmd/godcr-terminal
go build ./cmd/godcr-web
```
NB: Using `go build ./cmd/...` does not store but discards the built binaries.

### CODE REVIEW

**Review by barisere — CHANGES_REQUESTED** [review]
The new `godcr-web` binary fails with the following error when run from `$GOPATH/bin`.
```
➜  godcr git:(pr412_separate_entrypoints) godcr-web 
The following wallets were found...
(1) /Users/barisere/Library/Application Support/Dcrwallet/mainnet
(2) /Users/barisere/Library/Application Support/Dcrwallet/testnet3
(C)reate a new wallet.
(R)estore wallet from seed.
Select the wallet to use for this session: 2
2019/09/03 23:48:23 error loading templates: open ../../web/views/error.html: no such file or directory
```
We may need to use a tool such as [Packr](https://github.com/gobuffalo/packr) to isolate the built binary from the filesystem. This will enable us ship only binaries for the web UI, just as it is for other interfaces.

**Review by barisere — CHANGES_REQUESTED** [review]
`godcr-nuklear` no longer launches the GUI as expected.
```
➜  godcr git:(pr412_separate_entrypoints) godcr-nuklear 
The following wallets were found...
(1) /Users/barisere/Library/Application Support/Dcrwallet/mainnet
(2) /Users/barisere/Library/Application Support/Dcrwallet/testnet3
(C)reate a new wallet.
(R)estore wallet from seed.
Select the wallet to use for this session: 2
```

File: README.md [hunk 1]
```diff
@@ -42,53 +42,60 @@ git clone https://github.com/raedahgroup/godcr %GOPATH%/src/github.com/raedahgro
...
+* Alternatively, you can build only the binary for the interface you intend to run.
+Run `go build ./cmd/godcr-{interface}` or `go install ./cmd/godcr-{interface}`.
+* If running the http interface (`godcr-web`), you'd need to first build the the http frontend: `cd web/static/app`
+and run `yarn install` then `yarn build`.
+You can get yarn from [here](https://yarnpkg.com/lang/en/docs/install/)
 
 **Note: Building on Windows**
-Exporting `GO111MODULE` directly in CLI does not work and it is recommended to trigger
-via a `.bat` file.
+Exporting `GO111MODULE` directly in CLI does not work and
```
Comment (itswisdomagain) [inline README.md:56]: Don't know who added this to the readme, doesn't sound very accurate. Will need someone running Windows OS to confirm this.

**Review by barisere — CHANGES_REQUESTED** [review]
`godcr-fyne` launches the GUI, then crashes on start because it cannot find its assets. Again, [Packr](https://github.com/gobuffalo/packr) can help us here.
```
➜  godcr git:(pr412_separate_entrypoints) godcr-fyne 
The following wallets were found...
(1) /Users/barisere/Library/Application Support/Dcrwallet/mainnet
(2) /Users/barisere/Library/Application Support/Dcrwallet/testnet3
(C)reate a new wallet.
(R)estore wallet from seed.
Select the wallet to use for this session: 2
2019/09/03 23:58:21 overview png file missing open ../../fyne/pages/png/overview.png: no such file or directory
```

File: README.md (hunk 1 above)
Comment (barisere) [reply README.md:56]: I've seen that elsewhere too. I think it's because of the difference in handling of environment variables between *nix and Windows.

**Comment by itswisdomagain** [comment]
@barisere the 3 issues you identified above are all related to the interfaces not being able to find the required assets. Related issue https://github.com/raedahgroup/godcr/issues/239.

File: cmd/godcr-cli/log.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// initLogRotator initializes the logging rotater to write logs to logFile and
+// create roll files in the same directory.  It must be called before the
+// package-global log rotater variables are used.
+func initLogRotator(logFile string) {
+	logDir, _ := filepath.Split(logFile)
+	err := os.MkdirAll(logDir, 0700)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to create log directory: %v\n", err)
```
Comment (Sirmorrison) [inline cmd/godcr-cli/log.go]: `Failed`

File: cmd/godcr-cli/log.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+		os.Exit(1)
+	}
+	r, err := rotator.New(logFile, 10*1024, false, 0)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to create file rotator: %v\n", err)
```
Comment (Sirmorrison) [inline cmd/godcr-cli/log.go]: `Failed`

File: cmd/godcr-fyne/fyne.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	if err := parseAndSetDebugLevels(appConfig.DebugLevel); err != nil {
+		err := fmt.Errorf("loadConfig: %s", err.Error())
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+		return
+	}
+
+	// check if user passed commands/options/args in non-cli (fyne) mode
+	if len(args) > 0 {
+		fmt.Fprintf(os.Stderr, "unexpected command or flag in %s mode: %s\n",
```
Comment (Sirmorrison) [inline cmd/godcr-fyne/fyne.go]: `Unexpected`

File: cmd/godcr-nuklear/log.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// initLogRotator initializes the logging rotater to write logs to logFile and
+// create roll files in the same directory.  It must be called before the
+// package-global log rotater variables are used.
+func initLogRotator(logFile string) {
+	logDir, _ := filepath.Split(logFile)
+	err := os.MkdirAll(logDir, 0700)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to create log directory: %v\n", err)
```
Comment (Sirmorrison) [inline cmd/godcr-nuklear/log.go]: `Failed`

File: cmd/godcr-nuklear/log.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+		os.Exit(1)
+	}
+	r, err := rotator.New(logFile, 10*1024, false, 0)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to create file rotator: %v\n", err)
```
Comment (Sirmorrison) [inline cmd/godcr-nuklear/log.go]: `Failed`

File: cmd/godcr-terminal/log.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// initLogRotator initializes the logging rotater to write logs to logFile and
+// create roll files in the same directory.  It must be called before the
+// package-global log rotater variables are used.
+func initLogRotator(logFile string) {
+	logDir, _ := filepath.Split(logFile)
+	err := os.MkdirAll(logDir, 0700)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to create log directory: %v\n", err)
```
Comment (Sirmorrison) [inline cmd/godcr-terminal/log.go]: `Failed`

File: cmd/godcr-terminal/log.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+		os.Exit(1)
+	}
+	r, err := rotator.New(logFile, 10*1024, false, 0)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to create file rotator: %v\n", err)
```
Comment (Sirmorrison) [inline cmd/godcr-terminal/log.go]: `Failed`

File: cmd/godcr-terminal/terminal.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+	if err := parseAndSetDebugLevels(appConfig.DebugLevel); err != nil {
+		err := fmt.Errorf("loadConfig: %s", err.Error())
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+		return
+	}
+
+	// check if user passed commands/options/args in non-cli (terminal) mode
+	if len(args) > 0 {
+		fmt.Fprintf(os.Stderr, "unexpected command or flag in %s mode: %s\n",
```
Comment (Sirmorrison) [inline cmd/godcr-terminal/terminal.go]: `Unexpected`

File: cmd/godcr-web/log.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// initLogRotator initializes the logging rotater to write logs to logFile and
+// create roll files in the same directory.  It must be called before the
+// package-global log rotater variables are used.
+func initLogRotator(logFile string) {
+	logDir, _ := filepath.Split(logFile)
+	err := os.MkdirAll(logDir, 0700)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to create log directory: %v\n", err)
```
Comment (Sirmorrison) [inline cmd/godcr-web/log.go]: same as above.

**Review by Sirmorrison — CHANGES_REQUESTED** [review]

**Comment by Sirmorrison** [comment]
image

unused import error on cli.

**Review by Sirmorrison — APPROVED** [review]

**Comment by Sirmorrison** [comment]
cli, terminal interface commands look good.

fyne, nuklear, web interface works but need to be in the interface directory. (fix is on the way)

**Pull request merged**